File: erpnext/production/report/production_summary/production_summary.js
// Copyright (c) 2016, Frappe Technologies Pvt. Ltd. and contributors
// For license information, please see license.txt

frappe.query_reports["Production Summary"] = {
	onload:function(query_report){
		if  (query_report.get_filter_values("periodicity").periodicity != 'Daily'){
			frappe.query_report.toggle_filter_display('fiscal_year', false);
			frappe.query_report.toggle_filter_display('from_date', true);
			frappe.query_report.toggle_filter_display('to_date', true);
		}else{
			frappe.query_report.toggle_filter_display('fiscal_year', true);
			frappe.query_report.toggle_filter_display('from_date', false);
			frappe.query_report.toggle_filter_display('to_date', false);
		}
		if  (query_report.get_filter_values("periodicity").periodicity == 'Yearly'){
			frappe.query_report.toggle_filter_display('to_fiscal_year', false);
		}else{
			frappe.query_report.toggle_filter_display('to_fiscal_year', true);
		}
		query_report.refresh()

	},
	"filters": [
		{
			"fieldname": "cost_center",
			"label": __("Cost Center"),
			"fieldtype": "Link",
			"options": "Cost Center"
		},
		{
			"fieldname": "branch",
			"label": __("Branch"),
			"fieldtype": "Link",
			"options": "Branch"
		},
		{
			"fieldname": "warehouse",
			"label": __("Warehouse"),
			"fieldtype": "Link",
			"options": "Warehouse"
		},
		{
			"fieldname": "to_warehouse",
			"label": __("Transfer to Warehouse"),
			"fieldtype": "Link",
			"options": "Warehouse"
		},
		{
			"fieldname": "location",
			"label": __("Location"),
			"fieldtype": "Link",
			"options": "Location"
		},
		{
			"fieldname": "filter_based_on",
			"label": __("Filter Base On"),
			"fieldtype": "Select",
			"options": ["Fiscal Year","Date Range"],
			"default":"Fiscal Year",
			"hidden":1
		},
		{
			"fieldname": "periodicity",
			"label": __("Periodicity"),
			"fieldtype": "Select",
			"options": [
				{ "value": "Daily", "label": __("Daily") },
				{ "value": "Monthly", "label": __("Monthly") },
				{ "value": "Quarterly", "label": __("Quarterly") },
				{ "value": "Half-Yearly", "label": __("Half-Yearly") },
				{ "value": "Yearly", "label": __("Yearly") }
			],
			"default": "Daily",
			"reqd": 1,
			"on_change":(query_report)=>{
				if  (query_report.get_filter_values("periodicity").periodicity != 'Daily'){
					frappe.query_report.toggle_filter_display('fiscal_year', false);
					frappe.query_report.toggle_filter_display('from_date', true);
					frappe.query_report.toggle_filter_display('to_date', true);
				}else{
					frappe.query_report.toggle_filter_display('fiscal_year', true);
					frappe.query_report.toggle_filter_display('from_date', false);
					frappe.query_report.toggle_filter_display('to_date', false);
				}
				if  (query_report.get_filter_values("periodicity").periodicity == 'Yearly'){
					frappe.query_report.toggle_filter_display('to_fiscal_year', false);
				}else{
					frappe.query_report.toggle_filter_display('to_fiscal_year', true);
				}
				query_report.refresh()
			}
		},
		{
			"fieldname": "fiscal_year",
			"label": __("Fiscal Year"),
			"fieldtype": "Link",
			"options": "Fiscal Year",
			"default": frappe.defaults.get_user_default("fiscal_year"),
			"reqd": 1
		},
		{
			"fieldname": "to_fiscal_year",
			"label": __("To Fiscal Year"),
			"fieldtype": "Link",
			"options": "Fiscal Year",
			"default": frappe.defaults.get_user_default("fiscal_year"),
		},
		{
			"fieldname":"from_date",
			"label": __("From Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.month_start()
		},
		{
			"fieldname":"to_date",
			"label": __("To Date"),
			"fieldtype": "Date",
			"default":frappe.datetime.get_today()
		},
		{
			"fieldname": "company",
			"label": __("Company"),
			"fieldtype": "Link",
			"options": "Company",
			"default": frappe.defaults.get_user_default("company"),
			"read_only":1
		}
	]
}

File: erpnext/production/report/production_summary/production_summary.py
# Copyright (c) 2013, Frappe Technologies Pvt. Ltd. and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
from frappe import _
from frappe.utils import (flt, getdate, get_first_day, get_last_day, add_months, add_days, formatdate)
from erpnext.accounts.report.financial_statements import (
	get_data,
	get_filtered_list_for_consolidated_report,
	get_period_list,
)
def execute(filters=None):
    if filters.periodicity == 'Daily':
        period_list = ""
    else: 
        if filters.periodicity == 'Yearly':
            from_fiscal_year = filters.fiscal_year,
            to_fiscal_year   = filters.to_fiscal_year
        else:
            from_fiscal_year = filters.fiscal_year,
            to_fiscal_year   = filters.fiscal_year
        period_list = get_period_list(
					from_fiscal_year    = from_fiscal_year,
					to_fiscal_year      = to_fiscal_year,
					period_start_date   = getdate(str(filters.fiscal_year + '-01-01')),
					period_end_date     = getdate(str(filters.fiscal_year + '-12-31')),
					filter_based_on     = filters.filter_based_on,
					periodicity         = filters.periodicity,
					company             =filters.company)
    
    columns = get_columns(filters, period_list)
    data = get_data(filters, period_list)
    return columns, data

def get_data(filters, period_list):
    data= []
    conditions = get_conditions(filters)

    if filters.periodicity == 'Daily':
        query1 = frappe.db.sql("""
                    SELECT  
                        ppi.item_code, ppi.item_name
                    FROM `tabProduction` p, `tabProduction Product Item` ppi
                    WHERE ppi.parent = p.name AND p.docstatus = 1 GROUP BY ppi.item_code
                """, as_dict=1)

        for i in query1:
            row = {}
            total = 0
            flag = 0
            
            for cur_date in get_dates(filters):
                query = frappe.db.sql("""
                    SELECT 
                        ppi.item_code, 
                        sum(ppi.qty) as qty
                    FROM `tabProduction` p, `tabProduction Product Item` ppi 
                    WHERE ppi.parent = p.name AND p.docstatus = 1
                    AND p.posting_date = '{from_date}' {cond}
                    GROUP BY ppi.item_code;
                """.format(from_date=cur_date, cond=conditions), as_dict=True)

                for q in query:
                    if q.item_code == i.item_code:
                        total += q.qty
                        row["date_" + str(formatdate(cur_date, 'yyyyMMdd'))] = q.qty
                        row['total'] = total
                        flag = 1
            
            if flag == 1:
                row['item_code'] = '<b>' + i.item_code + '</b>'
                row['item_name'] = '<b>' + i.item_name + '</b>'
                data.append(row)
    else:
        query1 = frappe.db.sql(
                """
                    SELECT  
                        ppi.item_code, ppi.item_name
                    FROM `tabProduction` p, `tabProduction Product Item` ppi
                    WHERE ppi.parent = p.name AND p.docstatus = 1 GROUP BY ppi.item_code
                """, as_dict=1)

        for i in query1:
            row = {}
            total = 0
            flag = 0
            
            for d in period_list:
                query = frappe.db.sql("""
                    SELECT 
                        ppi.item_code, 
                        sum(ppi.qty) as qty
                    FROM `tabProduction` p, `tabProduction Product Item` ppi 
                    WHERE ppi.parent = p.name AND p.docstatus = 1
                    AND p.posting_date BETWEEN '{from_date}' AND '{to_date}' {cond}
                    GROUP BY ppi.item_code;
                """.format(from_date=d.from_date, to_date=d.to_date, cond=conditions), as_dict=True)

                for q in query:
                    if q.item_code == i.item_code:
                        total += q.qty
                        row[d.key] = q.qty
                        row['total'] = total
                        flag = 1
            
            if flag == 1:
                row['item_code'] = '<b>' + i.item_code + '</b>'
                row['item_name'] = '<b>' + i.item_name + '</b>'
                data.append(row)
    return data

def get_columns(filters , period_list):
    columns = [
        {
            "fieldname": "item_code",
            "label": _("Material Code"),
            "fieldtype": "Link",
            "options": "Item",
            "width": 80
        },
        {
            "fieldname": "item_name",
            "label": _("Material Name"),
            "fieldtype": "Data",
            "width": 200
        }
    ]

    if filters.periodicity == 'Daily':
        if filters.from_date and filters.to_date:
            if getdate(filters.from_date) > getdate(filters.to_date):
                frappe.throw("From Date should not be before To Date")

            for cur_date in get_dates(filters):
                columns.append({
                    "fieldname": "date_" + str(formatdate(cur_date, 'yyyyMMdd')),
                    "label": str(formatdate(cur_date, 'dd/MM/yyyy')),
                    "fieldtype": "Float"
                })
    else:
        for period in period_list:
            columns.append({
                "fieldname": period.key,
                "label": period.label,
                "fieldtype": "Data",
                "width": 120
            })

    columns.append({
        "fieldname": "total",
        "label": _("Grand Total"),
        "fieldtype": "Float",
        "width": 120
    })
    return columns

def get_dates(filters):
    date_list = []
    current_date = getdate(filters.from_date)
    while True:
        date_list.append(current_date)
        current_date = add_days(current_date, 1)
        if current_date > getdate(filters.to_date):
            break
    return date_list

def get_conditions(filters):
    conditions = ""
    if filters.get("cost_center"):
        conditions += " and p.cost_center = '{}'".format(filters.cost_center)
    if filters.get("branch"):
        conditions += " and p.branch = '{}'".format(filters.branch)
    if filters.get("warehouse"):
        conditions += " and p.warehouse = '{}'".format(filters.warehouse)
    if filters.get("to_warehouse"):
        conditions += " and p.to_warehouse = '{}'".format(filters.to_warehouse)
    if filters.get("location"):
        conditions += " and p.location = '{}'".format(filters.location)
    return conditions

